jdaviz/core/astrowidgets_api.py

```diff
@@ -402,105 +402,139 @@ def _validate_marker_name(self, marker_name):
+            if hasattr(in_reg, "to_pixel"):  # Sky region
+                if wcs is None:
+                    warnings.warn(
+                        f"Failed to mark this region, skipping: {in_reg}", AstropyUserWarning)
+                    continue
+                reg = in_reg.to_pixel(wcs)
             else:
-                ra_str = 'Right Ascension'
-                dec_str = 'Declination'
-            with jglue.data_collection.delay_link_manager_update():
-                jglue.data_collection[marker_name] = t_glue
-                jglue.add_link(t_glue, 'ra', image, ra_str)
-                jglue.add_link(t_glue, 'dec', image, dec_str)
-        else:
-            t_glue = Data(marker_name, **table[x_colname, y_colname])
-            with jglue.data_collection.delay_link_manager_update():
-                jglue.data_collection[marker_name] = t_glue
-                jglue.add_link(t_glue, x_colname, image, image.pixel_component_ids[1].label)
-                jglue.add_link(t_glue, y_colname, image, image.pixel_component_ids[0].label)
-
-        try:
-            self.add_data(t_glue)
-        except Exception as e:  # pragma: no cover
-            self.session.hub.broadcast(SnackbarMessage(
-                f"Failed to add markers '{marker_name}': {repr(e)}",
-                color="warning", sender=self))
-        else:
-            # Only can set alpha and color using self.add_data(), so brute force here instead.
-            # https://github.com/glue-viz/glue/issues/2201
-            for lyr in self.state.layers:
-                if isinstance(lyr, ScatterLayerState) and lyr.layer.label == marker_name:
-                    for key, val in self.marker.items():
-                        setattr(lyr, {'markersize': 'size'}.get(key, key), val)
-                    break
-
-            self.jdaviz_app.set_data_visibility(self.reference_id, marker_name,
-                                                visible=True, replace=False)
+                reg = in_reg
+
+            if isinstance(reg, PolygonPixelRegion):
+                x_coords = reg.vertices.x
+                y_coords = reg.vertices.y
+
+            # bqplot marker does not respect image pixel sizes, so need to render as polygon.
+            elif isinstance(reg, RectanglePixelRegion):
+                reg = reg.to_polygon()
+                x_coords = reg.vertices.x
+                y_coords = reg.vertices.y
+            elif isinstance(reg, (CirclePixelRegion, EllipsePixelRegion,
+                                  CircleAnnulusPixelRegion)):
+                # Astropy regions cannot convert these to polygons natively, so use glue.
+                x_coords, y_coords = regions2roi(reg).to_polygon()
 
-            self._marktags.add(marker_name)
+            else:
+                warnings.warn(
+                    f"Failed to mark this region, skipping: {in_reg}", AstropyUserWarning)
+                continue
+
+            new_marks.append(FootprintOverlay(

```

Comment:
this might conflict with logic in the footprints plugin which assumes all marks from this class are created by it (but if that is the case - we can just create a base/subclass or something)

Reply:
Theoretically it might be possible to refactor Footprints to use Astrowidgets API directly; then when you add a Footprint, you are really adding a marker but in a different way. But I didn't want to go down that rabbit hole yet.

Reply:
And for the Markers plugin, the concept is very similar to the "click to mark" portion of Astrowidgets API. Tying that in for image viewer is also theoretically possible, but I am not sure what to do with the spectrum stuff.